File: padroesComportamentais/padrao-de-projeto-iterator/java/src/main/java/Carrinho.java
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class Carrinho implements Iterable<Produto>{
    private List<Produto> produtos = new ArrayList<Produto>();

    public Carrinho adicionaProduto(Produto produto) {
        this.produtos.add(produto);
        return this;
    }

    @Override
    public Iterator<Produto> iterator() {
        return this.produtos.iterator();
    }
}

File: padroesComportamentais/padrao-de-projeto-iterator/php/src/Carrinho.php
<?php

namespace Iterator\Php;

use ArrayIterator;
use Ds\Vector;
use IteratorAggregate;

class Carrinho implements IteratorAggregate
{
    /** @var Produto[] */
    private Vector $produtos;

    public function __construct()
    {
        $this->produtos = new Vector();
    }

    public function adicionarProduto(Produto $produto): Carrinho
    {
        $this->produtos->push($produto);
        return $this;
    }

    public function getIterator(): \Iterator
    {
        return new CarrinhoIterator($this->produtos, false);
    }
}